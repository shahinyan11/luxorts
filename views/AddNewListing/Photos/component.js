import { Form } from 'antd';
import { FormikProvider } from 'formik';
import { FormattedMessage } from 'react-intl';

import AddNewListingLayout from 'views/layouts/AddNewListing';
import UploadPhotosField from 'views/shared/form/UploadPhotosField';

import useContainer, { getInitialProps } from './hook';

const Photos = () => {
  const { formik, showSkeleton, stepStatus, isLoading, onBackRoute, onRemovePhotoCallback } =
    useContainer();

  return (
    <AddNewListingLayout
      onNext={formik.handleSubmit}
      onBackRoute={onBackRoute}
      title={{ id: 'shared.photos' }}
      showSkeleton={showSkeleton}
      stepStatus={stepStatus}
      isLoading={isLoading}
      formik={formik}
    >
      <div className="new-listing__container new-listing__container--wide mb-8">
        <h1 className="new-listing__title mb-8">
          <FormattedMessage id="shared.uploadListingPhotos" />
        </h1>
        <p className="new-listing__description mb-32">
          <FormattedMessage id="listing.photos.imageFormat" />
        </p>
        <Form layout="vertical" size="large">
          <FormikProvider value={formik}>
            <UploadPhotosField name="listingPhotos" onRemove={onRemovePhotoCallback} />
          </FormikProvider>
        </Form>
      </div>
    </AddNewListingLayout>
  );
};

Photos.getInitialProps = getInitialProps;

export default Photos;
