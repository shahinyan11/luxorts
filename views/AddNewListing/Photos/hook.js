import { useFormik } from 'formik';
import { useSelector } from 'react-redux';
import { useRouter } from 'next/router';
import { useCallback, useEffect } from 'react';

import ROUTES from 'constants/routes';

import { createRoute } from 'utils/createRouteHelpers';
import isPresent from 'utils/isPresent';
import checkUserVerification from 'utils/auth/checkUserVerification';
import checkListingCompleteness from 'utils/listing/checkListingCompleteness';

import useParametricSelector from 'hooks/useParametricSelector';
import useFormSubmit from 'hooks/useFormSubmit';

import {
  fetchListing,
  fetchListingPhotos,
  updateListingPhotos,
} from 'state/concepts/listings/actions';
import { listingPhotosSelector, listingSelector } from 'state/concepts/listings/selectors';
import {
  fetchListingEndpoint,
  fetchListingPhotosEndpoint,
  updateListingPhotosEndpoint,
} from 'state/concepts/listings/endpoints';
import { loadingSelectorCreator } from 'state/data/selectors';
import { dataDeleteEntity } from 'state/data/actions';
import validationSchema from 'lib/yupLocalised/scheme/listing/photos';

function useContainer() {
  const router = useRouter();
  const onSubmit = useFormSubmit(updateListingPhotos);

  const listingId = router.query?.listingId;

  // endpoints
  const { endpoint } = fetchListingEndpoint(listingId);
  const { endpoint: fetchEndpoint } = fetchListingPhotosEndpoint(listingId);
  const { endpoint: updateEndpoint } = updateListingPhotosEndpoint(listingId);

  // selectors
  const listing = useSelector((state) => listingSelector(state, listingId));
  const listingPhotos = useSelector(listingPhotosSelector);
  const isFetching = useParametricSelector(loadingSelectorCreator, endpoint);
  const isFetchingPhotos = useParametricSelector(loadingSelectorCreator, fetchEndpoint);
  const isLoading = useParametricSelector(loadingSelectorCreator, updateEndpoint);

  // loaders and skeletons conditions
  const showSkeleton = isFetching || !isPresent(listing) || isFetchingPhotos;

  // prepare data from listing
  const stepStatus = listing?.applicationStep;
  const onBackRoute = createRoute({
    pathname: ROUTES.ADD_NEW_LISTING.SERVICES.PATH,
    id: listingId,
  });

  /**
   * Formik initialization
   */
  const formik = useFormik({
    enableReinitialize: false,
    initialValues: {
      listingPhotos,
      listingId,
      photosForDeleting: [],
      redirectRoute: createRoute({
        pathname: ROUTES.ADD_NEW_LISTING.DESCRIPTION.PATH,
        id: listingId,
      }),
    },
    validationSchema,
    onSubmit,
  });

  /**
   * On remove photo callback
   * @param photo {object}
   * @type {(function(*): void)|*}
   */
  const onRemovePhotoCallback = useCallback(
    (photo) => {
      if (!isPresent(photo.file)) {
        const { photosForDeleting } = formik.values;

        formik.setFieldValue('photosForDeleting', [
          ...photosForDeleting,
          { ...photo, delete: true },
        ]);
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [formik.values.photosForDeleting],
  );

  /**
   * Handle update formik listingPhotos value
   */
  const handleUpdateFormikListingPhotos = () => {
    if (!isLoading) {
      formik.resetForm({
        values: {
          ...formik.values,
          listingPhotos,
        },
      });
    }
  };

  /**
   * Lifecycle
   */
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(handleUpdateFormikListingPhotos, [listingPhotos, isLoading]);

  return {
    formik,
    showSkeleton,
    stepStatus,
    isLoading,
    onBackRoute,
    onRemovePhotoCallback,
    handleUpdateFormikListingPhotos,
  };
}

export const getInitialProps = async (ctx) => {
  const listingId = ctx.query?.listingId;
  const isListingIdPresent = isPresent(listingId);
  const isServer = Boolean(ctx.req);
  const isUserVerified = checkUserVerification(ctx);

  if (!isUserVerified) {
    return;
  }

  ctx.store.dispatch(dataDeleteEntity('listingPhoto'));

  if (isListingIdPresent) {
    ctx.store.dispatch(fetchListing(listingId));
  }

  if (isServer) {
    await ctx.store.logicMiddleware.whenComplete(() => {
      checkListingCompleteness(ctx);
    });
  }

  // fetch after check listing completeness
  if (isListingIdPresent) {
    ctx.store.dispatch(fetchListingPhotos(listingId));
  }
};

export default useContainer;
