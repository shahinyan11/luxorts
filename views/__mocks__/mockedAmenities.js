export default {
  'Amina Heaney': {
    '5e066ad7-4194-4bd4-95ec-0d9715f8159b': {
      id: '5e066ad7-4194-4bd4-95ec-0d9715f8159b',
      description: 'Maiores porro nostrum dolorum sunt.',
      name: 'Nolan Kertzmann',
      status: 'active',
      category: null,
      checked: false,
    },
  },
  'Torrie Sawayn': {
    'c277033f-1683-4d2d-a409-d00b0bbe8201': {
      id: 'c277033f-1683-4d2d-a409-d00b0bbe8201',
      description: 'Nemo magnam dolores quam voluptatibus distinctio unde.',
      name: 'Stephine Rohan',
      status: 'active',
      category: null,
      checked: false,
    },
  },
};
