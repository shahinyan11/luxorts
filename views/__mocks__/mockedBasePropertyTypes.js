export default [
  {
    value: '51e35121-c27e-46fc-9feb-b51dab2490ee',
    label: 'repudiandae',
    description: 'description repudiandae',
  },
  {
    value: '995c5714-185d-4085-8512-690903334f6e',
    label: 'molestias',
    description: 'description molestias',
  },
];
