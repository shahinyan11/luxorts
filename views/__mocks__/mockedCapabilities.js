export default {
  APPROVED: {
    '11da7909-9f4c-45d6-a2a0-8a105b067ed2': {
      id: '11da7909-9f4c-45d6-a2a0-8a105b067ed2',
      description: 'Vel commodi ad incidunt repellendus a molestias delectus nam.',
      name: 'Tonya Wunsch',
      status: 'active',
      checked: false,
    },
    'b84c898e-2176-4ac8-9dfa-54186ccef6cb': {
      id: 'b84c898e-2176-4ac8-9dfa-54186ccef6cb',
      description: 'Maxime nostrum ex iste velit nulla incidunt.',
      name: 'Edwin Greenholt',
      status: 'active',
      checked: false,
    },
  },
};
