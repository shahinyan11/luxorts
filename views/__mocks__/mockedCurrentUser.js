export default {
  id: '5c5d814c-8e06-433d-b006-4119311585bd',
  email: 'evhenii.hryhoriev+10@rubygarage.org',
  publicId: 'c4toxHVi',
  verifiedStatus: 'unverified',
  userType: 'traveller_account',
  userProfile: {
    id: '52080e19-c5e0-45bb-af58-edbbe70e78fa',
    about: null,
    address: null,
    avatarUrls: {
      original: '/avatar-img-url',
    },
    dateOfBirth: '1994-01-01',
    drivingLicenseFirstName: 'Evhenii',
    drivingLicenseLastName: 'Test',
    firstName: 'Evhenii',
    gender: null,
    lastName: 'Test',
    phoneNumber: '+299999999',
    street: 'street',
    city: 'city',
    state: 'state',
    zipCode: 'zipCode',
    country: 'country',
  },
  userInvitation: [
    {
      id: '52080e19-c5e0-45bb-af58-edbbe70e78fa',
      created_at: '2022-06-23T06:32:55.588Z',
      email: '12@mail.ru',
      last_sent_at: '2022-06-23T06:32:55.474Z',
      resendable: true,
      status: 'pending',
      user_account_id: '7211ccc2-1fca-4799-b1e1-4005e9fc42be',
    },
  ],
};
