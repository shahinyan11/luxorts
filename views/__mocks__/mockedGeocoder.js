export default [
  {
    streetNumber: '240',
    street: '240 North 7th Avenue',
    city: 'Phoenix',
    state: 'Arizona',
    country: 'United States',
    zipCode: '85007',
    fullAddress: '240 N 7th Ave, Phoenix, AZ 85007, USA',
    longitude: -112.0825682,
    latitude: 33.4507026,
  },
];
