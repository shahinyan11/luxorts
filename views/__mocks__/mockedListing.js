export default {
  id: '0645baf3-945d-4122-8a0b-5ab3e95a9b3e',
  applicationStep: 'accomodation',
  status: 'in_draft',
  userAccount: { id: '6cbf2b68-4a26-41b5-bcd5-161cc37e1420', type: 'hostAccount' },
  listingPropertyType: {
    id: 'ea19f127-bc89-4799-bdbb-346126ac3845',
    typeOfPlace: 'private_place',
    propertyType: {
      id: '51e35121-c27e-46fc-9feb-b51dab2490ee',
      name: 'repudiandae',
      description: 'description repudiandae',
    },
  },
  listingLocation: {
    id: '2b4032ec-a469-4489-b7dc-536d8f9552eb',
    country: 'United States',
    city: 'Phoenix',
    state: 'Arizona',
    street: 'West Maryland Avenue',
    apartmentNumber: '12',
    zipCode: '12345',
    longitude: '-73.97447621784085',
    latitude: '40.764468117826794',
  },
  listingDescription: null,
  listingPricing: {
    pricePerDay: 64,
  },
};
