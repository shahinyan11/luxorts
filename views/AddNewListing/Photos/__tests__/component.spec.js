import { shallow } from 'enzyme';

import { formik } from '__mocks__/formik';

import Photos from '../component';

const mockedHookData = {
  formik,
  onBack: jest.fn(),
  handleSubmit: jest.fn(),
  onRemovePhotoCallback: jest.fn(),
};
jest.mock('../hook', () => jest.fn(() => mockedHookData));

describe('Photos component', () => {
  const component = shallow(<Photos />);

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
