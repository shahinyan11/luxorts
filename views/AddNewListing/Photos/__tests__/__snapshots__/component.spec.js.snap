// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photos component matches snapshot 1`] = `
<AddNewListingLayout
  formik={
    Object {
      "dirty": true,
      "handleSubmit": [MockFunction],
      "isValid": true,
      "resetForm": [MockFunction],
      "setFieldTouched": [MockFunction],
      "setFieldValue": [MockFunction],
      "setValues": [MockFunction],
      "validateForm": [MockFunction],
      "values": Object {},
    }
  }
  isDirty={false}
  isLoading={false}
  onNext={[MockFunction]}
  showSkeleton={false}
  stepStatus="initial"
  title={
    Object {
      "id": "shared.photos",
    }
  }
>
  <div
    className="new-listing__container new-listing__container--wide mb-8"
  >
    <h1
      className="new-listing__title mb-8"
    >
      <Memo(FormattedMessage)
        id="shared.uploadListingPhotos"
      />
    </h1>
    <p
      className="new-listing__description mb-32"
    >
      <Memo(FormattedMessage)
        id="listing.photos.imageFormat"
      />
    </p>
    <ForwardRef(InternalForm)
      layout="vertical"
      size="large"
    >
      <ContextProvider
        value={
          Object {
            "dirty": true,
            "handleSubmit": [MockFunction],
            "isValid": true,
            "resetForm": [MockFunction],
            "setFieldTouched": [MockFunction],
            "setFieldValue": [MockFunction],
            "setValues": [MockFunction],
            "validateForm": [MockFunction],
            "values": Object {},
          }
        }
      >
        <UploadPhotosField
          name="listingPhotos"
          onRemove={[MockFunction]}
        />
      </ContextProvider>
    </ForwardRef(InternalForm)>
  </div>
</AddNewListingLayout>
`;
